### DISCUSSION TITLE
VRS digest serialization and SequenceReference

### DISCUSSION THREAD

**Discussion by ahwagner** [ga4gh/vrs#465] [Ideas]
In VRS 1.x, if an object was GA4GH-identifiable, the digest of that object is used in place as a reference to that object in nested structures, simplifying the input serialization for computed identifier calculation.

In VRS 2.x, this remains the same, but now we have other Value Objects with integrated descriptive attributes that are dropped for serialization. We also have a "special case" of SequenceReference, which is non-identifiable, but passes through a digest of a RefGet sequence object for digest serialization.

Here is what that looks like from a serialization perspective:
![VRS 2 x Digest Process - current model](https://github.com/ga4gh/vrs/assets/811065/cf98be37-6bf4-472d-8a49-6fc3529a35d9)

I would like to consider treating all value objects in VRS consistently, by _always_ computing/retrieving and returning a digest for serialization. We can maintain the "passthrough" special case for SequenceReference to avoid computing a secondary digest, but this would provide consistency in implementation by always returning digests for any Value Object in VRS. All Value Objects already allow for _digest_ to be stored, so "compute digest" may be implemented as a 1-time operation per value object. The revised process would look like this:
![VRS 2 x Digest Process - proposed model](https://github.com/ga4gh/vrs/assets/811065/8293f835-fde3-49b5-b986-de1c3a979fa2)

This has nice implications for other complex "for purpose" value objects like GenotypeMember, which could also be compacted into digest references.

A second option is to go the other way: instead non-GA4GH-identifiable objects are serialized but not digested for parent object serialization, and the digest keys for SequenceReference are limited to type and refgetAccession. It would look like this:
![VRS 2 x Digest Process - proposed model 2](https://github.com/ga4gh/vrs/assets/811065/f4e55c45-0a7b-4165-88c6-af43f3dc7182)

My preference is this last model, since there are no special cases under this model. 

@reece @larrybabb @korikuzma @theferrit32 @tnavatar @toneillbroad @andreasprlic thoughts?

**Comment by larrybabb** [comment]
My preference is also the last model. I only see the costs and not the value in creating digests for objects that are not identifiable. We will occassionally show compacted representation of higher order classes (like Allele) that contain only the nested identifiable object(s) (SequenceLocation) as a digest. We would never do this for these other structures that are NOT identifiable.

**Reply by ahwagner** [reply]
One point to consider here is how to handle unordered arrays of non-GA4GH-identifiable classes used in Value Objects. The only place this is currently an issue is `GenotypeMember`. Since `Genotype` is not part of our `2.0` effort, I propose we leave that issue alone until it needs to be addressed.
